### DISCUSSION TITLE
Rotation extension

### DISCUSSION THREAD

**Discussion by mcencini** [imr-framework/pypulseq#184] [Ideas]
Hi all,
thanks for this amazing work! I wanted to discuss an undocumented Pulseq extension (i.e, rotation), I hope this is the appropriate place for it.

**Motivation**
Pulseq format does a fantastic job in keeping the size of the seqfile small when the sequence mostly consists of repetitions of scaled waveforms (which is true for most Cartesian imaging). However, I feel the base format is not so efficient when it comes to Non Cartesian sequences (pulseq/pulseq#31). Luckily, many Non Cartesian trajectories can be efficiently described as rotated versions of a single waveform (see, for instance, [this](https://onlinelibrary.wiley.com/doi/full/10.1002/mrm.29194)), potentially enabling a significant reduction of memory footprint and faster loading times, even in older scanner platforms.

Apparently, [C++ implementation](https://github.com/pulseq/pulseq/tree/master/src) of Pulseq already defines a **Rotation event** extension, representing the rotation of the gradient matrix for each block. This extension, as far as I know, has never been exposed to MATLAB or Python codebases, perhaps because the main Siemens interpreters do not support it? I am not a Siemens user myself, so I cannot verify this.

**Proposal**
I exposed the rotation extension in my own [PyPulseq fork](https://github.com/mcencini/pypulseq/tree/rot_ext), since the GE interpreter potentially supports this kind of waveform manipulation already (@jfnielsen). I also adapted the plotting tools and helpers routines extracting gradients, waveforms and so on. Before opening a pull request, I would like to know if that sounds like a good idea. Personally, I do not see this as a too much intrusive modification, as per [Pulseq specification](https://pulseq.github.io/specification.pdf) extensions are explicitly designed to add new features without having to go through a major release. As an example, code for sequences aiming to be compatible with different interpreters (supporting rotation or not), could contain this kind of code snippet for the gradient rotation:

```python
if use_rot_ext:
    # Create rotation event
    rot = pp.make_rotation(create_rot_matrix(phi, axis="z"))

    # Add gradient and rotation event
    seq.add_block(grad, rot)
else:
    # Add rotated gradient
     seq.add_block(*pp.rotate(grad, angle=phi, axis="z"))
```
Thanks again!
Matteo

**Comment by btasdelen** [comment]
@namgyunlee This maybe of interest to you.

**Reply by namgyunlee** [reply]
@btasdelen Thank you for letting me know! Hope this feature is realized in the Siemens interpreter as well :)

**Reply by WSHoge** [reply]
fyi.  A matlab version of this extension idea is available here: https://github.com/WSHoge/pulseq/tree/rotext

**Reply by mdtisdall** [reply]
@mcencini, thanks for your work on this! Are you aware of any plans to integrate this back into the official release of pypulseq? I'm sure I'm not the only one who'd like to be able to emit .seq files with the rotation extension.

**Reply by mcencini** [reply]
Hi! I plan to open a PR once v1.5.0 is published - rotation extension is included in MATLAB 1.5.1 branch, so I wanted to follow it. In the meantime, I included it in my own [fork](https://github.com/mcencini/pypulseq), if you want to try it.

**Reply by mdtisdall** [reply]
Awesome, thanks! Will test my sequence generated using the extension via your fork with the latest release of the Siemens interpreter and report back.
